File: kotlin-server/net/src/main/kotlin/ccr/net/bitstream/BitPacker.kt
package ccr.net.bitstream

// C++: cBitPacker in wwbitpack/BitPacker.h/.cpp
// Ported from the Jani 02-14-2002 optimized rewrite (MSB-first bit ordering).
// Note: old version used LSB-first — the two are not compatible. We use the new version.
//
// MAX_BUFFER_SIZE = 548 (MTU 576 - 20 IP header - 8 UDP header)
const val MAX_BUFFER_SIZE = 548

open class BitPacker {
    val buffer = ByteArray(MAX_BUFFER_SIZE)

    // Independent read and write positions, both in bits
    var bitWritePosition: Int = 0
        private set
    var bitReadPosition: Int = 0
        private set

    fun reset() {
        bitWritePosition = 0
        bitReadPosition = 0
        buffer.fill(0)
    }

    fun flush() {
        bitReadPosition = bitWritePosition
    }

    val isFlushed: Boolean get() = bitReadPosition == bitWritePosition

    // C++: Set_Bit_Write_Position — used when constructing a packet from received raw bytes
    fun setBitWritePosition(position: Int) {
        require(position <= MAX_BUFFER_SIZE * 8) { "position $position exceeds buffer capacity" }
        bitWritePosition = position
    }

    // C++: Add_Bits — MSB-first encoding (Jani's 02-14-2002 optimized rewrite)
    fun addBits(value: Long, numBits: Int) {
        require(numBits in 1..32) { "numBits must be 1..32, was $numBits" }
        require(bitWritePosition + numBits <= MAX_BUFFER_SIZE * 8) {
            "write overflow: pos=$bitWritePosition + bits=$numBits > ${MAX_BUFFER_SIZE * 8}"
        }

        var byteNum = bitWritePosition ushr 3
        val bitOffset = bitWritePosition and 0x7
        bitWritePosition += numBits

        // Shift value to MSB of a 32-bit word
        var remaining = numBits
        var v = (value.toInt()) shl (32 - numBits)

        // Fill the remaining bits of the current byte first
        if (bitOffset != 0) {
            val bitCount = minOf(8 - bitOffset, remaining)
            val bitValue = v.ushr(24 + bitOffset).toByte()
            buffer[byteNum] = (buffer[byteNum].toInt() or bitValue.toInt()).toByte()
            byteNum++
            v = v shl bitCount
            remaining -= bitCount
        }

        // Write remaining full bytes and partial last byte
        while (remaining > 0) {
            buffer[byteNum] = v.ushr(24).toByte()
            byteNum++
            v = v shl 8
            remaining -= 8
        }
    }

    // C++: Get_Bits — reads numBits bits MSB-first
    fun getBits(numBits: Int): Long {
        require(numBits in 1..32) { "numBits must be 1..32, was $numBits" }
        require(bitReadPosition + numBits <= bitWritePosition) {
            "read overflow: readPos=$bitReadPosition + bits=$numBits > writePos=$bitWritePosition"
        }

        val readLen = numBits
        var byteNum = bitReadPosition / 8
        val bitOffset = bitReadPosition and 0x7
        bitReadPosition += numBits

        var bitCount = 8 - bitOffset
        if (bitCount > numBits) bitCount = numBits

        var value = ((buffer[byteNum++].toInt() and 0xFF shl bitOffset) and 0xFF).toLong() shl 24
        var remaining = numBits - bitCount

        var shift = 24 - bitCount
        while (shift > 0) {
            value = value or ((buffer[byteNum++].toInt() and 0xFF).toLong() shl shift)
            shift -= 8
            remaining -= 8
        }
        if (remaining > 0) {
            value = value or ((buffer[byteNum].toInt() and 0xFF).toLong() ushr (-shift))
        }

        value = value ushr (32 - readLen)
        return value and 0xFFFFFFFFL
    }
}

File: original/Code/wwbitpack/BitPacker.cpp
//
// Filename:     bitpacker.cpp
// Project:      wwbitpack.lib
// Author:       Tom Spencer-Smith
// Date:         June 1998
// Description:  Minimal bit encoding
//

#include "bitpacker.h"

#include <string.h>	// for memset


//-----------------------------------------------------------------------------
//cBitPacker::cBitPacker(UINT buffer_size) :
cBitPacker::cBitPacker() :
	//BufferSize(buffer_size),
	BitWritePosition(0),
	BitReadPosition(0)
{
	//assert(BufferSize > 0);

	//Buffer = new BYTE[BufferSize];
	//assert(Buffer != NULL);
	//memset(Buffer, 0, BufferSize);
	memset(Buffer, 0, MAX_BUFFER_SIZE);
}

//-----------------------------------------------------------------------------
cBitPacker::~cBitPacker() 
{
	//delete [] Buffer;
}

//-----------------------------------------------------------------------------
cBitPacker& cBitPacker::operator=(const cBitPacker& rhs)
{
	//assert(BufferSize == rhs.BufferSize);

	//memcpy(Buffer, rhs.Buffer, rhs.BufferSize);
	memcpy(Buffer, rhs.Buffer, MAX_BUFFER_SIZE);
	BitReadPosition		= rhs.BitReadPosition;
	BitWritePosition		= rhs.BitWritePosition;

   return * this;
}

//-----------------------------------------------------------------------------
//
// This method needs optimization
//
// 02-14-2002 Jani: Optimized the code somewhat. Note that the old code reverted
// the bit order and the new one doesn't, so the versions are not compatible.
// If you use optimized Add_Bits() you need to also use optimize Get_Bits().
//

void cBitPacker::Add_Bits(ULONG value, UINT num_bits)
{
	//
	// N.B. Presently you cannot use this class with an atomic type of more 
	// than 4 bytes, such as a double. Hopefully you would be using a float 
	// instead anyway.
	//
#if 0	// Old version

	ULONG mask = 1 << (num_bits - 1);
	while (mask > 0) {

		//assert(BitWritePosition < BufferSize * 8);

		UINT byte_num = BitWritePosition / 8;
		UINT bit_offset = BitWritePosition % 8;
		bool bit_value = (value & mask) != 0;
		Buffer[byte_num] |= bit_value << bit_offset;

		BitWritePosition++;

		mask >>= 1;
	}

#else	// New faster version

	// Verify that we're not writing over buffer

	// Fill the remaining bits of the write byte first
	UINT byte_num = BitWritePosition >> 3;
	UINT bit_offset = BitWritePosition & 0x7;
	BitWritePosition+=num_bits;		// Advance the write position

	// If write buffer is not byte aligned, write the remaining bits first
	value <<= 32-num_bits;
	if (bit_offset) {
		UINT bit_count = 8 - bit_offset;
		if (bit_count>num_bits) bit_count=num_bits;

		ULONG bit_value = value;
		value <<= bit_count;					// Remove the copied bits
		num_bits -= bit_count;
		bit_value >>= (24+bit_offset);
		Buffer[byte_num++] |= bit_value;
	}

	// Write the rest of the data as bytes
	if (num_bits>8) {
		for (unsigned a=0;a<num_bits;a+=8) {
			Buffer[byte_num++]=(unsigned char)(value>>24);
			value<<=8;
		}
	}
	else {
		Buffer[byte_num]=(unsigned char)(value>>24);
	}
#endif
}

//-----------------------------------------------------------------------------
//
// This method needs optimization
// 02-14-2002 Jani: Optimized. See Add_Bits() for notes.
//
void cBitPacker::Get_Bits(ULONG & value, UINT num_bits)
{
#if 0	// Old version

	value = 0;
	for (int bit = num_bits - 1; bit >= 0; bit--) {

		//assert(BitReadPosition < BufferSize * 8);
		UINT byte_num = BitReadPosition / 8;
		UINT bit_offset = BitReadPosition % 8;
		bool b = (Buffer[byte_num] & (1 << bit_offset)) != 0;

		value += (b << bit);	

		BitReadPosition++;
	}
#else // New faster version

	// Verify that we're not reading over buffer or write pointer

	UINT read_len=num_bits;
	UINT byte_num = BitReadPosition / 8;
	UINT bit_offset = BitReadPosition % 8;
	BitReadPosition += num_bits;

	UINT bit_count = 8 - bit_offset;
	if (bit_count>num_bits) bit_count=num_bits;
	value = (ULONG(Buffer[byte_num++]) << (bit_offset+24));
	num_bits-=bit_count;

	int shift;
	for (shift=24-bit_count;shift>0;shift-=8,num_bits-=8) value|=unsigned(Buffer[byte_num++]) << shift;
	if (num_bits>0) value|=Buffer[byte_num++]>>(-shift);

	value >>= 32-read_len;
#endif
}

//-----------------------------------------------------------------------------
//
// This method is only for use by a packet class when data is received.
//

void cBitPacker::Set_Bit_Write_Position(UINT position)
{
	//assert(position <= BufferSize * 8);
	BitWritePosition = position;
}










/*
//-----------------------------------------------------------------------------
void cBitPacker::Increment_Bit_Position(int num_bits)
{

	for (int i = 0; i < num_bits; i++) {
		Advance_Bit_Position();
		NumBits++;
	}
}

//-----------------------------------------------------------------------------
UINT cBitPacker::Get_Compressed_Size_Bytes() const 
{
	return (int) ceil(BitWritePosition / 8.0f);
}

//-----------------------------------------------------------------------------
inline void cBitPacker::Advance_Bit_Position()
{
	BitWritePosition++;
	
	//
	// If the following assert hits then our buffer is not large enough.
	// We can advance BitWritePosition one bit past the end of the buffer, but
	// we cannot write there.
	//
	//assert(BitWritePosition < BufferSize * 8);
}

*/


